File: backend/app/services/blockchain_service.py
"""
Blockchain service for processing blockchain data
Handles API calls to blockchain.info and data transformation
"""
import httpx
import asyncio
from typing import List, Dict, Any
from datetime import datetime
from app.models.schemas import (
    AddressResponse,
    Transaction,
    GraphNode,
    GraphLink,
    GraphData,
)
from app.services.rate_limiter import wait_for_rate_limit


BLOCKCHAIN_API_BASE = "https://blockchain.info"
DEFAULT_TIMEOUT = 30.0  # sec


async def fetch_address_details(
    address: str,
    limit: int = 50,
    offset: int = 0,
    timeout: float = DEFAULT_TIMEOUT,
) -> AddressResponse:
    """
    Fetch address details from blockchain.info API
    
    Args:
        address: Bitcoin address to fetch
        limit: Number of transactions to fetch
        offset: Number of transactions to skip
        timeout: Request timeout in seconds
    
    Returns:
        AddressResponse with transactions
    """
    url = f"{BLOCKCHAIN_API_BASE}/rawaddr/{address}?limit={limit}&offset={offset}"
    
    # Always wait to respect API rate limits (1 request every 10 seconds)
    await wait_for_rate_limit("blockchain_api")
    
    async with httpx.AsyncClient(timeout=timeout) as client:
        try:
            response = await client.get(url, headers={"Accept": "application/json"})
            
            # If we still get rate limited (429), wait longer before retry
            if response.status_code == 429:
                # Wait additional 10 seconds before retry (API may still be rate limiting)
                await asyncio.sleep(10)
                await wait_for_rate_limit("blockchain_api")  # Wait another 10 seconds
                response = await client.get(url, headers={"Accept": "application/json"})
            
            response.raise_for_status()
            
            data = response.json()
            return AddressResponse(**data)
        except httpx.HTTPStatusError as e:
            # Re-raise with more context
            if e.response.status_code == 429:
                raise httpx.HTTPStatusError(
                    "Rate limited by blockchain.info API. Please wait a few minutes before trying again.",
                    request=e.request,
                    response=e.response,
                )
            elif e.response.status_code == 503:
                raise httpx.HTTPStatusError(
                    "Blockchain.info API is temporarily unavailable. This may be due to rate limiting or maintenance.",
                    request=e.request,
                    response=e.response,
                )
            raise


def convert_transactions_to_graph(
    address: str,
    transactions: List[Transaction],
) -> GraphData:
    """
    Convert blockchain transactions to graph format (nodes and links)
    
    Args:
        address: The central address being investigated
        transactions: List of transaction objects from blockchain.info API
    
    Returns:
        GraphData with nodes and links for graph visualization
    """
    nodes: Dict[str, GraphNode] = {}
    links: List[GraphLink] = []
    
    # Add the central address as a node
    nodes[address] = GraphNode(
        id=address,
        label=f"{address[:8]}...{address[-8:]}",
    )
    
    for tx in transactions:
        tx_hash = tx.hash
        tx_time = tx.time
        
        # INBOUND: Process inputs (money coming FROM other addresses TO our target address)
        for input_item in tx.inputs:
            prev_out = input_item.prev_out
            if not prev_out:
                continue
                
            source_addr = prev_out.get("addr")
            value = prev_out.get("value", 0)
            
            if not source_addr:
                continue
            
            # Check if this transaction sends money TO our target address
            target_address_in_outputs = any(
                output.addr == address for output in tx.out
            )
            
            # Only create link if money goes TO target address
            if target_address_in_outputs:
                # Add source node if it doesn't exist
                if source_addr not in nodes:
                    nodes[source_addr] = GraphNode(
                        id=source_addr,
                        label=f"{source_addr[:8]}...{source_addr[-8:]}",
                    )
                
                # Create INBOUND link: source_addr -> address (target)
                links.append(GraphLink(
                    source=source_addr,
                    target=address,
                    value=value,
                    txHash=tx_hash,
                    timestamp=tx_time,
                ))
        
        # OUTBOUND: Process outputs (money going FROM our target address TO other addresses)
        for output in tx.out:
            dest_addr = output.addr
            value = output.value
            
            if not dest_addr or dest_addr == address:
                continue
            
            # Check if this transaction sends money FROM our target address
            source_address_in_inputs = any(
                inp.prev_out and inp.prev_out.get("addr") == address
                for inp in tx.inputs
            )
            
            # Only create link if money comes FROM target address
            if source_address_in_inputs:
                # Add destination node if it doesn't exist
                if dest_addr not in nodes:
                    nodes[dest_addr] = GraphNode(
                        id=dest_addr,
                        label=f"{dest_addr[:8]}...{dest_addr[-8:]}",
                    )
                
                # Create OUTBOUND link: address (source) -> dest_addr
                links.append(GraphLink(
                    source=address,
                    target=dest_addr,
                    value=value,
                    txHash=tx_hash,
                    timestamp=tx_time,
                ))
    
    return GraphData(
        nodes=list(nodes.values()),
        links=links,
    )


def merge_graph_data(
    existing: GraphData,
    new_data: GraphData,
) -> GraphData:
    """
    Merge two graph data structures, avoiding duplicates
    
    Args:
        existing: Existing graph data
        new_data: New graph data to merge
    
    Returns:
        Merged graph data
    """
    node_map: Dict[str, GraphNode] = {}
    link_map: Dict[str, GraphLink] = {}
    
    # Process existing nodes and links
    for node in existing.nodes:
        node_map[node.id] = node
    
    for link in existing.links:
        key = f"{link.source}-{link.target}-{link.txHash}"
        link_map[key] = link
    
    # Add new nodes (if they don't exist)
    for node in new_data.nodes:
        if node.id not in node_map:
            node_map[node.id] = node
    
    # Add new links (if they don't exist)
    for link in new_data.links:
        key = f"{link.source}-{link.target}-{link.txHash}"
        if key not in link_map:
            link_map[key] = link
    
    return GraphData(
        nodes=list(node_map.values()),
        links=list(link_map.values()),
    )

File: src/lib/services/blockchain-service.ts
/**
 * Blockchain Service
 * High-level service layer for blockchain operations
 */
import { clearApiLog, fetchAddressDetails, fetchAddressGraph, getApiLog } from '../api/backend-client';

import type { AddressResponse, ApiLogEntry, GraphData, GraphLink, GraphNode } from '../types/blockchain';

/**
 * Merge two graph data structures, avoiding duplicates
 */
export function mergeGraphData(existing: GraphData, newData: GraphData): GraphData {
  const nodeMap = new Map<string, GraphNode>();
  const linkMap = new Map<string, GraphLink>();

  // Process existing nodes and links
  existing.nodes.forEach((node) => nodeMap.set(node.id, node));
  existing.links.forEach((link) => {
    const key = `${link.source}-${link.target}-${link.txHash}`;
    linkMap.set(key, link);
  });

  // Add new nodes (if they don't exist)
  newData.nodes.forEach((node) => {
    if (!nodeMap.has(node.id)) {
      nodeMap.set(node.id, node);
    }
  });

  // Add new links (if they don't exist)
  newData.links.forEach((link) => {
    const key = `${link.source}-${link.target}-${link.txHash}`;
    if (!linkMap.has(key)) {
      linkMap.set(key, link);
    }
  });

  return {
    nodes: Array.from(nodeMap.values()),
    links: Array.from(linkMap.values()),
  };
}

export function formatSatoshisToBTC(satoshis: number): string {
  return (satoshis / 100_000_000).toFixed(8);
}

export function formatTimestamp(timestamp: number): string {
  return new Date(timestamp * 1000).toLocaleString();
}

export function isValidBitcoinAddress(address: string): boolean {
  if (!address || address.length < 26 || address.length > 100) {
    return false;
  }

  if (/^[13][a-km-zA-HJ-NP-Z1-9]{25,34}$/.test(address)) {
    return true;
  }

  if (/^bc1[a-z0-9]{39,59}$/i.test(address)) {
    return true;
  }

  if (/^bc1p[a-z0-9]{58}$/i.test(address)) {
    return true;
  }

  return false;
}

export class BlockchainService {
  static async getAddressDetails(address: string, limit = 50, offset = 0): Promise<AddressResponse> {
    if (!isValidBitcoinAddress(address)) {
      throw new Error('Invalid Bitcoin address format');
    }
    // Check if this is a Taproot address (bc1p...) - not supported by blockchain.info
    if (address.startsWith('bc1p')) {
      throw new Error('Taproot addresses (bc1p...) are not supported by the blockchain API');
    }

    // Check if the address is too long (additional safety check)
    if (address.length > 90) {
      throw new Error('This address format is not supported by the blockchain API');
    }

    return fetchAddressDetails(address, limit, offset);
  }

  static async getAddressGraph(address: string, limit = 50, offset = 0): Promise<GraphData> {
    if (!isValidBitcoinAddress(address)) {
      throw new Error('Invalid Bitcoin address format');
    }

    return fetchAddressGraph(address, limit, offset);
  }

  static getApiLog(): ApiLogEntry[] {
    return getApiLog();
  }

  static clearApiLog(): void {
    clearApiLog();
  }
}
